### ISSUE TITLE
Stopping a title is not reliable: ShutdownTitle hangs in OSSchedulerEnd or Latte_Stop

### ISSUE THREAD

**Issue by WizzardSK** [WizzardSK/cemu-libretro#15]
Split out of #7, where it is the remaining blocker rather than the subject.

`CafeSystem::ShutdownTitle()` does not reliably finish. When it works it takes about **65 ms**; when it does not, it never returns at all, and the libretro core is left taking its `_exit(0)` escape hatch — which is why nobody has noticed before: the process was ending either way, so "the title stopped" and "the title did not stop" looked identical from outside.

Measured on aarch64 (postmarketOS, Adreno 618, Vulkan/Turnip, RetroArch flatpak), NSMBU, closing content after ~40 s of play. Two distinct places:

### 1. `OSSchedulerEnd()`, joining core 1

```
[OSScheduler] OSSchedulerEnd woke cores, joining threads...
[OSScheduler] OSSchedulerEnd joining thread 0
[OSScheduler] Core 0 exited idle loop, thread exiting
[OSScheduler] Core 2 exited idle loop, thread exiting
[OSScheduler] OSSchedulerEnd thread 0 joined
[OSScheduler] OSSchedulerEnd joining thread 1      ← nothing after this
```

Cores 0 and 2 see the shutdown and leave; core 1 does not. Polling the scheduler's own counters once a second while it hangs says why:

```
core0  heartbeat=14    idleEnter=49427   idleWake=49427
core1  heartbeat=2316  idleEnter=0       idleWake=0
core2  heartbeat=7     idleEnter=305420  idleWake=305420
```

Core 1's instruction heartbeat is **frozen** — it is not executing guest code — and it has **never once** entered the scheduler idle loop in the whole session. That is where `sSchedulerActive` is read, so the wake `OSSchedulerEnd` sends (incrementing every core's run-queue count) has nothing to wake: core 1 is parked in a host call made from guest code and never comes back to the loop that would notice. The last guest IP differs run to run (`0x029b4fe8`, `0x029cfd88`), so it is not one particular place in the title.

This is not a deadlock on `sSchedulerStateMtx`, which `OSSchedulerEnd` holds across the joins — that mutex appears exactly twice in the file, in `OSSchedulerBegin` and `OSSchedulerEnd`, and no core thread takes it.

### 2. `Latte_Stop()`, with the GPU thread still alive

```
[OSScheduler] OSSchedulerEnd done
[LatteThread] Latte_Stop begin running=1 finishedInit=1      ← nothing after this
```

`LatteThread` was still in the process nine minutes later, waiting on a futex, along with the IOSU threads. RetroArch had already unloaded the core and could not terminate.

### Why it matters beyond a clean exit

The bounded shutdown in `dbb6443` exists so a title's open files are closed, and therefore flushed, before the process leaves. Every time this hangs, that flush does not happen. Saves are not actually at risk today — `366fa88` flushes explicitly on `FSAProcessCmd_flushQuota`, during play and independent of shutdown — but a title that never calls it has no second line of defence.

It also blocks the rest of #7: with the title stopped, the renderer teardown now runs to completion and RetroArch survives (`content closed, core unloaded`, exit 0). Without it, the core still has to end the process.

### Note on frequency

Rate is not a fixed property — it moved with an unrelated change. Stopping the input manager *before* the title (rather than after) turned it from occasional into six hangs out of six, which is what identified the ordering; that is fixed in d90ba4e. With the correct order it is intermittent again, roughly one run in two here, and the two failure sites above are both still reachable.

**Comment by WizzardSK** [comment]
**Correction to the issue body: site 2 is wrong.** I claimed the shutdown also hangs in `Latte_Stop()`. It does not — that run had no phase tracing left in it, so "the last line is `Latte_Stop begin`" only meant nothing logged afterwards; the steps after it (`snd_core::reset`, `OSAlarm_Shutdown`, `_GX2DriverReset`, the unmounts) log nothing at all. I read more into it than it said.

With every step of `ShutdownTitle` traced, the hang is site 1, every time:

```
[ShutdownTitle] begin (+0ms)      ← and nothing more
```

i.e. inside `coreinit::OSSchedulerEnd()`, at the `join()` on core 1 — matching the counters in the body.

**New data.** Sampling every thread of the process from `/proc` while it hangs (three samples, three seconds apart, all identical):

* Every thread is asleep. 24 of them on `futex` (syscall 98), four on `poll`, one on `nanosleep` — that last one is the core's own bounded wait. **Nothing is spinning**, so core 1 is not stuck in a busy loop; it is blocked on a futex nobody signals.
* `LatteThread` is alive and also in `futex_wait` — idle, not working.

So the shape is a lost wakeup or a two-sided wait between the guest thread parked on core 1 and something that will only be produced while the emulator still runs. That fits the timing dependence: it is about what core 1 happens to be doing at the moment the content is closed, not about which title or which run.

**Two things in `LatteThread.cpp` that are worth fixing regardless, neither of which is this hang:**

1. In libretro mode `Latte_Stop()` **detaches** the GPU thread rather than joining it:
   ```cpp
   #ifdef ENABLE_LIBRETRO
       if (sLatteThread.joinable()) sLatteThread.detach();
   #else
       sLatteThread.join();
   #endif
   ```
   So `ShutdownTitle()` can return while the GPU thread is still running — and d90ba4e destroys the renderer immediately afterwards, on the device that thread draws through. The comment says it is there to avoid a deadlock on blocking GL calls; on the Vulkan path we want the opposite, a bounded join.

2. `Latte_Start()` holds `sLatteThreadStateMutex` across the whole wait for `sLatteThreadFinishedInit`, and `Latte_Stop()` needs that same mutex. An initialisation that never completes deadlocks the stop. Not what happened here — the log showed `finishedInit=1` — but it is loaded.

**Comment by WizzardSK** [comment]
Fixed in f672a06 / 2bfeb1c / a62a126.

## Case 1: `OSSchedulerEnd()`, joining core 1

The idle loop has an exit for every core except the one that hangs. `__OSThreadCoreIdle` checks `sSchedulerActive` and switches back to the scheduler fiber only in its `else` branch — the one a *secondary* core takes after its run-queue semaphore is released. The main core takes the `isMainCore` branch, which calls `__OSCheckSystemEvents()` and loops. There is nothing in it that ever ends the loop.

So the main core's only way out was `__OSThreadSwitchToNext`, which runs only if a guest thread happens to be scheduled on that core when the scheduler is stopped. Whether stopping a title worked depended on what the title was doing at that moment — which is exactly the "intermittent" in the title of this issue.

One correction to the numbers in the report above: `idleEnter=0` for core 1 is not evidence of anything. `sSchedulerIdleWaitEnterCount` is only incremented in the secondary-core branch, so the main core never counts there by construction.

The fix gives the main core the same check the other cores have.

## Case 2: `Latte_Stop()`

Not a hang — `Latte_Stop` returned fine. In libretro builds it *detached* the GPU thread unconditionally, so `ShutdownTitle` reported success while a live thread went on rendering through the frontend's Vulkan device. What was left of it showed up as the crash at exit in `LatteShaderCache_Load` / `RendererShaderVk::CreateVkShaderModule` with a null device, and as `LatteThread` still in the process afterwards.

It now waits for the thread to finish its own teardown and joins it, keeps the detach as a bounded fallback, and — the part that matters — reports that fallback. A detached GPU thread means the title is only half stopped, and `retro_unload_game` already knows what to do with a title that did not stop.

Two things had to be fixed before that wait could ever succeed:

- The frame gate (#12) parks the GPU thread outside a frame, and parking is something that thread does for itself at a command boundary. Waiting for a parked thread to park always timed out — that is the `GPU thread did not park before the context went away` warning. The gate is now held open for the length of the pause handshake.
- An exiting frontend destroys its graphics context **before** it unloads the core. Everything Cemu does to tear its renderer down goes through that device, so doing it anyway faults inside the driver. When the context is already gone the renderer is now left alone.

## Measurements

`retroarch --max-frames=1200` on NSMBU turns out to be a deterministic way to exercise the unload path, which is what made this measurable at all.

| build | clean stops |
|---|---|
| before the frame gate | 3 / 3 |
| with the frame gate, before this fix | 3 / 6 |
| after this fix | 6 / 6 |

The gate did not cause the bug, it exposed it: parking the cores at a fixed point in the frame lands in the bad case far more often than free-running does.

## Still open, separately

The frontend's *exit* after the core is unloaded is still not clean — threads that outlive the library (IOSU services, a detached GPU thread) end in `terminate called without an active exception` at `dlclose`. Filed as a follow-up; it is a different problem from stopping a title, and it does not put save data at risk, which this one did.
